homeassistant/components/recorder/purge.py

```diff
@@ -101,3 +65,78 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
     except SQLAlchemyError as err:
         _LOGGER.warning("Error purging history: %s", err)
     return True
+
+
+def _select_event_ids_to_purge(session: Session, purge_before: datetime) -> list:
+    """Return a list of event ids to purge."""
+    events = (
+        session.query(Events.event_id)
+        .filter(Events.time_fired < purge_before)
+        .limit(MAX_ROWS_TO_PURGE)
+        .all()
+    )
+    _LOGGER.debug("Selected %s event ids to remove", len(events))
+    return [event.event_id for event in events]
+
+
+def _select_state_ids_to_purge(
+    session: Session, purge_before: datetime, event_ids: list
+) -> list:
+    """Return a list of state ids to purge."""
+    if not event_ids:
+        return []
+    states = (
+        session.query(States.state_id)
+        .filter(States.last_updated < purge_before)
+        .filter(States.event_id.in_(event_ids))
+        .all()
+    )
+    _LOGGER.debug("Selected %s state ids to remove", len(states))
+    return [state.state_id for state in states]
+
+
+def _disconnect_states_about_to_be_purged(session: Session, state_ids: list) -> None:
+    # Update old_state_id to NULL before deleting to ensure
+    # the delete does not fail due to a foreign key constraint
+    # since some databases (MSSQL) cannot do the ON DELETE SET NULL
+    # for us.
+    disconnected_rows = (
+        session.query(States)
+        .filter(States.old_state_id.in_(state_ids))
+        .update({"old_state_id": None}, synchronize_session=False)
+    )
+    _LOGGER.debug("Updated %s states to remove old_state_id", disconnected_rows)
+
+
+def _purge_state_ids(session: Session, state_ids: list) -> None:
+    """Delete by state id."""
+    deleted_rows = (
+        session.query(States)
+        .filter(States.state_id.in_(state_ids))
+        .delete(synchronize_session=False)
+    )
+    _LOGGER.debug("Deleted %s states", deleted_rows)
+
+
+def _purge_event_ids(session: Session, event_ids: list) -> None:
+    """Delete by event id."""
+    deleted_rows = (
+        session.query(Events)
+        .filter(Events.event_id.in_(event_ids))
```

Comment:
I added it to the `old_state_id` since that wasn't on the primary key.  Since this one is on the primary key, I think the query optimizer is always going to remove it.

Reply:
Sounds good. It could always be added later if we notice that it takes to long (maybe if some DBMS doesn't optimize it that way).